pkg/controller/registry/resolver/steps.go

```diff
@@ -132,40 +137,50 @@ func NewStepResourceFromBundle(bundle *api.Bundle, namespace, replaces, catalogS
 	annos[projection.PropertiesAnnotationKey] = anno
 	csv.SetAnnotations(annos)
 
-	step, err := NewStepResourceFromObject(csv, catalogSourceName, catalogSourceNamespace)
+	csvStep, err := NewStepResourceFromObject(csv, catalogSourceName, catalogSourceNamespace)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
-	steps := []v1alpha1.StepResource{step}
+	steps := []v1alpha1.StepResource{}
+	namespaces := []string{}
 
 	for _, object := range bundle.Object {
 		dec := yaml.NewYAMLOrJSONDecoder(strings.NewReader(object), 10)
 		unst := &unstructured.Unstructured{}
 		if err := dec.Decode(unst); err != nil {
-			return nil, err
+			return nil, nil, err
 		}
-
+		namespaces = append(namespaces, unst.GetNamespace())
 		if unst.GetObjectKind().GroupVersionKind().Kind == v1alpha1.ClusterServiceVersionKind {
+			// Adding the CSV step only here, although it was computed earlier, to keep the namespace and step orders aligned.
+			steps = append(steps, csvStep)
 			continue
 		}
 
 		step, err := NewStepResourceFromObject(unst, catalogSourceName, catalogSourceNamespace)
 		if err != nil {
-			return nil, err
+			return nil, nil, err
 		}
 		steps = append(steps, step)
 	}
 
 	operatorServiceAccountSteps, err := NewServiceAccountStepResources(csv, catalogSourceName, catalogSourceNamespace)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 	steps = append(steps, operatorServiceAccountSteps...)
-	return steps, nil
+	saNamespaces := []string{}
```

Comment:
I'm a bit unclear for why we need the namespace here. The catalogsource namespace is used because that where the bundle is unpacked but using it for identifier is really a safe choice? What happens if multiple operators have the same resources and all in the same catalogsource?

Reply:
This is a bit of an edge case and I am open to suggestions. The original intent of the enhancement is to allow operator installation when APIs of "optional resources" are not available.
ServiceAccounts are core resources and as such their API will always be available.
Wanting to make things consistent I kept what was populated into namespace for them. Another option would be to keep the namespace empty. Do you think something else would make more sense?